internal/serialization/default_compact_reader.go

```diff
@@ -0,0 +1,987 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package serialization
+
+import (
+	"fmt"
+	"time"
+
+	ihzerrors "github.com/hazelcast/hazelcast-go-client/internal/hzerrors"
+	pserialization "github.com/hazelcast/hazelcast-go-client/serialization"
+	"github.com/hazelcast/hazelcast-go-client/types"
+)
+
+const (
+	NULL_ARRAY_LENGTH         = -1
+	NULL_OFFSET               = -1
+	BYTE_OFFSET_READER_RANGE  = 127 - (-128)
+	SHORT_OFFSET_READER_RANGE = 32767 - (-32768)
+)
+
+type Reader func(*ObjectDataInput) interface{}
+type SliceReader func(*ObjectDataInput) []interface{}
+type SliceConstructor func(int32) interface{}
+
+var BYTE_OFFSET_READER = ByteOffsetReader{}
+var SHORT_OFFSET_READER = ShortOffsetReader{}
+var INT_OFFSET_READER = IntOffsetReader{}
+
+type OffsetReader interface {
+	getOffset(input *ObjectDataInput, variableOffsetsPos int32, index int32) int32
+}
+
+type ByteOffsetReader struct{}
+
+func (ByteOffsetReader) getOffset(inp *ObjectDataInput, variableOffsetsPos int32, index int32) int32 {
+	offset := inp.ReadByteAtPosition(variableOffsetsPos + index)
+	if offset == 0xFF {
+		return NULL_OFFSET
+	}
+	return int32(offset)
+}
+
+type ShortOffsetReader struct{}
+
+func (ShortOffsetReader) getOffset(inp *ObjectDataInput, variableOffsetsPos int32, index int32) int32 {
+	offset := inp.ReadInt16AtPosition(variableOffsetsPos + (index * Int16SizeInBytes))
+	return int32(offset)
+}
+
+type IntOffsetReader struct{}
+
+func (IntOffsetReader) getOffset(inp *ObjectDataInput, variableOffsetsPos int32, index int32) int32 {
+	return inp.ReadInt32AtPosition(variableOffsetsPos + (index * Int32SizeInBytes))
+}
+
+type DefaultCompactReader struct {
```

Comment:
No. I'll align them.

Reply:
Sorry, there is a reason. I can't do all of them pointer receiver. If I do that DefaultCompactReader is not implementing the CompactReader interface. How can I solve this? I want to make all pointer receiver because DefaultCompactReader is a large struct.

Reply:
Returning a pointer to the `DefaultCompactReader` from `NewDefaultCompactReader` should let us use pointer receivers

Reply:
Also, the static check printed an error just above this comment. Can you take a look at the padding issue?

Reply:
Fixed the fieldalignment problems.

Reply:
> Returning a pointer to the DefaultCompactReader from NewDefaultCompactReader should let us use pointer receivers
fixed in e514811